src/service/document-info-impl.js

```diff
@@ -84,11 +92,21 @@ export class DocInfo {
     const pageViewId = getPageViewId(ampdoc.win);
     const linkRels = getLinkRels(ampdoc.win.document);
     const metaTags = getMetaTags(ampdoc.win.document);
+    let replaceUrl = null;
 
     return this.info_ = {
       /** @return {string} */
       get sourceUrl() {
-        return getSourceUrl(ampdoc.getUrl());
+        // Ads served via the proxy may specify a "replaceUrl" to override the
+        // source URL.
+        if (allowReplaceUrl(ampdoc)) {
+          // The replace URL parameter takes precedence over the viewer
+          // messaging flow (the value is subsequently cached).
+          replaceUrl = replaceUrl ||
+              getReplaceUrlFromParameter(ampdoc) ||
+              Services.viewerForDoc(ampdoc).getReplaceUrl();
+        }
+        return replaceUrl || getSourceUrl(ampdoc.getUrl());
```

Comment:
shall we cache the return value from `getSourceUrl` as well?

Reply:
discussed offline, the sourceUrl was intentionally not cached. 
the reason needs to be revisited, but should not block this PR.